Notifications: Display the link in the subject line.

dragotin opened the pull request:
Added a new data called linkText. If the link in the activity JSON has a space in it, the part before that is considered the link text. If it does not contain a space, the word [link] is used.

Note that the idea to separate the link and the link text  with a space needed because the original specification does not contain a link text, just the link. I am not seeing a downside as long as spaces in the real link are encoded with %20.

This is may be a solution for #6236

dragotin commented:
DISCLAIMER: I did not have a ownCloud available that can do notifications, so this is PR is untested. Thus label "DO NOT MERGE YET" with the kind request to review/comment and maybe test.

ogoffart commented on src/gui/servernotificationhandler.cpp:
```diff
@@ -81,7 +81,16 @@ void ServerNotificationHandler::slotNotificationsReceived(const QJsonDocument &j
+            // is rendered as a link text
+            if( s.contains(QChar(' '))) {
+                const QStringList li = s.split(QChar(' '));
+                a._link = QUrl::fromEncoded(li.at(1).toLocal8Bit());
```
ToLatin1 or toUtf8.

toLocal8bit is for path on the filesystem, or things that we want to expose in stderr

There can only be one space?

ogoffart commented on src/gui/notificationwidget.cpp:
```diff
@@ -44,6 +44,16 @@ void NotificationWidget::setActivity(const Activity &activity)
+    if( !activity._link.isEmpty() ) {
+        // append a link to the message, if that is empty, to subject
+        subject.append(QString("&nbsp;<a href=\"%1\">%2</a>")
+                       .arg(activity._link.toString(QUrl::FullyEncoded))
```
use arg(... , ...)  with two arguments.   In general this is better because it works even if the first string contains a %1 or a %2.

Also, please use escapeXml  to the _linkText

ogoffart commented on src/gui/servernotificationhandler.cpp:
```diff
@@ -81,7 +81,16 @@ void ServerNotificationHandler::slotNotificationsReceived(const QJsonDocument &j
+                a._linkText = li.at(0);
+            } else {
+                a._link = QUrl(s);
+                a._linkText = QString("[%1]").arg(tr("link"));
```
(nitpick:  i'd use tr("[link]"))

ogoffart commented on src/gui/notificationwidget.cpp:
```diff
@@ -44,6 +44,16 @@ void NotificationWidget::setActivity(const Activity &activity)
     _ui._subjectLabel->setVisible(!activity._subject.isEmpty());
     _ui._messageLabel->setVisible(!activity._message.isEmpty());
 
+    QString subject = activity._subject;
```
needs to be xml escaped.

ogoffart commented:
Why not a new button next to "Close" to open the link?

guruz commented:
@dragotin nice PR! When do you want to continue on it?

> Why not a new button next to "Close" to open the link?

If there is a long link text, this might look odd. But if you just want to have a button "Open" then I think it's fine..

dragotin commented:
Fixed all the outstanding things (I hope) and added a unit test for splitting of the url from the link text. Thanks @ogoffart for review.

I do not like the idea of a button, because a link should not be a button, but a link. Also please note that it is already possible to add additional buttons that trigger a link via the notification specification, see here: https://github.com/owncloud/client/blob/master/src/gui/notificationwidget.cpp#L72

ogoffart commented on src/gui/activitydata.h:
```diff
@@ -57,6 +57,7 @@ class Activity
     QString _message;
     QString _file;
     QUrl _link;
+    QString _linkText;
```
Maybe document which field are html escaped.

Personally, i'd keep all these string unescaped.  and escape them right before putting them in some html text.

ogoffart commented on src/gui/activitydata.h:
```diff
@@ -69,6 +70,23 @@ class Activity
+            // if there is a real whitespace in the link, the part before the space
+            // is rendered as a link text
+            int charPos = s.lastIndexOf(QChar(' '));
+            if( charPos > -1 ) {
```
(style:  we don't put spaces inside the parentheses, the space is before.  You could run git-clang-format, to fix all style issues.)

ogoffart commented on src/gui/notificationwidget.cpp:
```diff
@@ -44,6 +44,20 @@ void NotificationWidget::setActivity(const Activity &activity)
+        }
+        subject.append( QString("&nbsp;<a href=\"%1\">%2</a>")
+                        .arg(activity._link.toString(QUrl::FullyEncoded),
+                             lText.toHtmlEscaped() ));
```
Am I wrong that you double html escape?

ogoffart commented on test/testactivitydata.cpp:
```diff
@@ -0,0 +1,68 @@
+  
+private slots:
+  
+    void testLinkSplitEmpty() {
```
this kind of test would be much shorter using a testLink_data()  thing

ogoffart replied:
_data tests, like for example TestFolder::testFolder or TestUtility::testSanitizeForFileName_data  or ...


And also QCOMPARE instead of QVERIFY

ogoffart commented on src/gui/notificationwidget.cpp:
```diff
@@ -44,6 +44,20 @@ void NotificationWidget::setActivity(const Activity &activity)
+        _ui._subjectLabel->setOpenExternalLinks(true);
+
+    }
     _ui._subjectLabel->setText(activity._subject);
```
you mean just "subject" here!

dragotin commented:
Thanks @ogoffart, I fixed stuff accordingly.

With the test I would need more advise ;-)

ogoffart commented:
I wanted to remove my "Request change" without "Approving" but apparently this is not possible.

dragotin commented:
Should be good now.

guruz commented:
Works for me, but let's wait if the server guys are OK with the format. I'll ping them.

<img width="769" alt="screen shot 2018-04-09 at 17 28 27" src="https://user-images.githubusercontent.com/959327/38506916-79957d32-3c1b-11e8-9130-6774e8c22638.png">

We can make it more beautiful later. Maybe the link should be in a newline?

DeepDiver1975 commented on src/gui/activitydata.h:
```diff
@@ -69,6 +71,23 @@ class Activity
+        if (!s.isEmpty()) {
+            // if there is a real whitespace in the link, the part before the space
+            // is rendered as a link text
+            int charPos = s.lastIndexOf(QChar(' '));
```
That's fishy ....

guruz commented:
Superseded by https://github.com/owncloud/client/pull/6453 after discussion on server issue about link text. Sorry @dragotin
